netsim/providers/__init__.py

```diff
@@ -142,52 +161,93 @@ def create_extra_files_mappings(
           module=self.provider)
         continue
 
-      out_folder = f"{self.provider}_files/{node.name}"
-      bind_dict[f"{out_folder}/{file}"] = mapping         # note: node_files directory is flat
+      bind_dict[f"{out_folder}/{file}"] = mapping
+
 
     node[self.provider][outkey] = filemaps.dict_to_mapping(bind_dict)
 
   def create_extra_files(
       self,
       node: Box,
       topology: Box,
-      inkey: str = 'config_templates',
       outkey: str = 'binds') -> None:
 
     binds = node.get(f'{self.provider}.{outkey}',None)
     if not binds:
       return
 
     sys_folder = str(_files.get_moddir())+"/"
-    out_folder = f"{self.provider}_files/{node.name}"
 
     bind_dict = filemaps.mapping_to_dict(binds)
+    # Process other files normally
+    node_data = {
+      **node.to_dict(),
+      'addressing': topology.addressing.to_dict(),
+      'hostvars': topology.nodes.to_dict(),
+      'hosts': topology.nodes.to_dict()
+    }
+    
+    # Performance: cache constants and use instance-level template cache
+    prefix = f"{self.provider}_files/"
+    prefix_len = len(prefix)
+
     for file,mapping in bind_dict.items():
-      if not out_folder in file:                  # Skip files that are not mapped into the temporary provider folder
+      # Only handle files placed under provider_files/
+      if not file.startswith(prefix):
         continue
-      file_name = file.replace(out_folder+"/","")
-      template_name = self.find_extra_template(node,file_name,topology)
-      if template_name:
-        node_data = node + { 'hostvars': topology.nodes, 
-                             'hosts': get_host_addresses(topology),
-                             'addressing': topology.addressing }  # Needed for subnet prefix
-        if '/' in file_name:                      # Create subdirectory in out_folder if needed
-          pathlib.Path(f"{out_folder}/{os.path.dirname(file_name)}").mkdir(parents=True,exist_ok=True)
-        try:
-          templates.write_template(
-            in_folder=os.path.dirname(template_name),
-            j2=os.path.basename(template_name),
-            data=node_data.to_dict(),
-            out_folder=out_folder, filename=file_name)
-        except Exception as ex:
-          log.fatal(
-            text=f"Error rendering {template_name} into {file_name}\n{strings.extra_data_printout(str(ex))}",
-            module=self.provider)
 
-        strings.print_colored_text('[MAPPED]  ','bright_cyan','Mapped ')
-        print(f"{out_folder}/{file_name} to {node.name}:{mapping} (from {template_name.replace(sys_folder,'')})")
+      # Derive the out_folder (either node-specific or shared per-device) and the relative filename
```

Comment:
Instead of most of this complex stuff, you could keep a cache of shared file paths and use the template name as the key to that cache. Then you just check whether the template is in the cache, and return shared filename if it is.

If the same template path that is marked as :shared renders into different content we have a bigger problem.

Also, while it's nice to have shared files in "shared/device/" folder, that's not absolutely necessary.